File: app/src/main/java/com/ruideraj/secretelephant/contacts/ContactAdapter.java
package com.ruideraj.secretelephant.contacts;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.ruideraj.secretelephant.R;

import java.util.List;

/**
 * Adapter for populating the list in ContactsListFragment
 */
public class ContactAdapter extends RecyclerView.Adapter {

    private List<Contact> mContacts;
    private ContactsViewModel mViewModel;
    private ContactClickListener mClickListener;

    public ContactAdapter(@NonNull ContactsViewModel viewModel,
                          ContactClickListener clickListener) {
        mViewModel = viewModel;
        mClickListener = clickListener;
    }

    @Override
    @NonNull
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.contact_list_item, parent, false);

        return new ViewHolder(view, mClickListener);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ViewHolder vh = (ViewHolder) holder;

        // Fill in contact info
        Contact contact = mContacts.get(position);

        vh.name.setText(contact.getName());
        vh.data.setText(contact.getData());

        // Show selected status with background color.
        if(mViewModel.selectedData.contains(contact.getData())) {
            vh.itemView.setBackgroundColor(ContextCompat.getColor(vh.itemView.getContext(),
                    R.color.bg_contact_selected));
        }
        else {
            vh.itemView.setBackgroundColor(ContextCompat.getColor(vh.itemView.getContext(),
                    R.color.bg_activity));
        }
    }

    @Override
    public int getItemCount() {
        return mContacts == null ? 0 : mContacts.size();
    }

    public void setData(List<Contact> data) {
        mContacts = data;
        notifyDataSetChanged();
    }

    public interface ContactClickListener {
        void onContactClick(int position);
    }

    private class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private ContactClickListener mClickListener;
        TextView name;
        TextView data;

        public ViewHolder(View view, ContactClickListener clickListener) {
            super(view);
            view.setOnClickListener(this);
            mClickListener = clickListener;
            name = view.findViewById(R.id.contactName);
            data = view.findViewById(R.id.contactData);
        }

        @Override
        public void onClick(View v) {
            int position = getAdapterPosition();
            if(position != RecyclerView.NO_POSITION) {
                mClickListener.onContactClick(position);
            }

        }
    }
}

File: app/src/main/java/com/ruideraj/secretelephant/contacts/ContactAdapter.kt
package com.ruideraj.secretelephant.contacts

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.ruideraj.secretelephant.R

class ContactAdapter(private val fragment: Fragment,
                     private val viewModel: ContactsViewModel,
                     private val listener: ContactClickListener)
    : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var contacts: List<Contact>? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.contact_list_item, parent, false)

        return ViewHolder(view, listener)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val contact = contacts!![position]

        val vh = holder as ViewHolder
        vh.name.text = contact.name
        vh.data.text = contact.data

        Glide.with(fragment)
                .load(contact.avatarUri)
                .placeholder(R.drawable.ic_contact_placeholder)
                .circleCrop()
                .into(vh.image)

        val colorId = if (viewModel.selectedContacts.contains(contact)) {
            R.color.bg_contact_selected
        } else {
            R.color.bg_activity
        }
        vh.itemView.setBackgroundColor(ContextCompat.getColor(vh.itemView.context, colorId))
    }

    override fun getItemCount() = contacts?.size ?: 0

    fun setData(data: List<Contact>) {
        contacts = data
        notifyDataSetChanged()
    }

    fun interface ContactClickListener {
        fun onContactClick(position: Int)
    }

    private class ViewHolder(itemView: View, listener: ContactClickListener) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private val clickListener = listener
        val name: TextView = itemView.findViewById(R.id.contactName)
        val data: TextView = itemView.findViewById(R.id.contactData)
        val image: ImageView = itemView.findViewById(R.id.contactImage)

        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            if (adapterPosition != RecyclerView.NO_POSITION) {
                clickListener.onContactClick(adapterPosition)
            }
        }
    }
}